source/extensions/transport_sockets/tls/ocsp/ocsp.cc

```diff
@@ -0,0 +1,307 @@
+  CBS_init(&cbs, reinterpret_cast<const uint8_t*>(der.c_str()), der.size());
+
+  auto resp = Asn1OcspUtility::parseOcspResponse(cbs);
+  if (CBS_len(&cbs) != 0) {
+    throw EnvoyException("Data contained more than a single OCSP response");
+  }
+
+  return resp;
+}
+
+void skipResponderId(CBS& cbs) {
+  // ResponderID ::= CHOICE {
+  //    byName               [1] Name,
+  //    byKey                [2] KeyHash
+  // }
+  //
+  // KeyHash ::= OCTET STRING -- SHA-1 hash of responder's public key
+  //    (excluding the tag and length fields)
+
+  if (Asn1Utility::isOptionalPresent(cbs, nullptr, CONS | CONT | 1) ||
+      Asn1Utility::isOptionalPresent(cbs, nullptr, CONS | CONT | 2)) {
+    return;
+  }
+
+  throw EnvoyException(absl::StrCat("Unknown choice for Responder ID: ", parseTag(cbs)));
+}
+
+} // namespace
+
+OcspResponse::OcspResponse(OcspResponseStatus status, ResponsePtr&& response)
+    : status_(status), response_(std::move(response)) {}
+
+BasicOcspResponse::BasicOcspResponse(ResponseData data, std::string signature_alg,
+                                     std::vector<uint8_t> signature)
+    : data_(data), signature_alg_(signature_alg), signature_(std::move(signature)) {}
+
+const std::string BasicOcspResponse::OID = "1.3.6.1.5.5.7.48.1.1";
+
+ResponseData::ResponseData(Envoy::SystemTime produced_at,
+                           std::vector<SingleResponse> single_responses)
+    : produced_at_(produced_at), single_responses_(std::move(single_responses)) {}
+
+SingleResponse::SingleResponse(CertId cert_id, CertStatus status, Envoy::SystemTime this_update,
+                               absl::optional<Envoy::SystemTime> next_update)
+    : cert_id_(cert_id), status_(status), this_update_(this_update), next_update_(next_update) {}
+
+CertId::CertId(std::string serial_number, std::string alg_oid, std::string issuer_name_hash,
+               std::string issuer_public_key_hash)
+    : serial_number_(serial_number), alg_oid_(alg_oid), issuer_name_hash_(issuer_name_hash),
+      issuer_public_key_hash_(issuer_public_key_hash) {}
+
+OcspResponseWrapper::OcspResponseWrapper(std::string der_response, TimeSource& time_source)
+    : raw_bytes_(std::move(der_response)), response_(readDerEncodedOcspResponse(raw_bytes_)),
+      time_source_(time_source) {
+
+  if (response_->response_ == nullptr) {
+    throw EnvoyException("OCSP response has no body");
+  }
+
+  // We only permit a 1:1 of certificate to response
+  if (response_->response_->getNumCerts() != 1) {
+    throw EnvoyException("OCSP Response must be for one certificate only");
+  }
+
+  auto& this_update = response_->response_->getThisUpdate();
+  if (time_source_.systemTime() < this_update) {
+    DateFormatter formatter("%E4Y%m%d%H%M%S");
+    throw EnvoyException(absl::StrCat("OCSP Response thisUpdate field is set in the future: ",
+                                      formatter.fromTime(this_update)));
+  }
+}
+
+// TODO(daniel-goldstein): This should also check the issuer
```

Comment:
Why can't this be implemented now?

Reply:
This can actually be implemented now fairly easily so I can lump that in.

Reply:
The only limitation here is the decision of what range of hash algorithms to accept. I found this set of hash algorithms based on BoringSSL digest.h
https://github.com/envoyproxy/envoy/blob/2a6a4b8d758b4bf4238cb6d2babdd3d83fbf45bc/source/extensions/common/crypto/utility_impl.cc#L81-L99

but the question remains of what to do when a hash algorithm outside this is used

Reply:
Sorry, I'm lacking some context on this. What do we need the hash function for?

Reply:
Sorry let me elaborate. To identify the certificate for which an OCSP response is for, it carries both the serial number of the certificate, the _hash_ of the certificate issuer's distinguished name, and the hash algorithm used to compute that hash. So if we want to verify that the name on the OCSP response matches the name on the certificate, we need to extract the name from the certificate, hash it according to the algorithm specified in the OCSP response, and check that it matches the hash in the OCSP response.

The above snippet is the supported list of hash algorithms I could find to accomplish this in Envoy. In a case where the specified algorithm in the OCSP response does not match one of these, do we just say that the certificate does not match the OCSP response?

Reply:
Oh, I'd just ignore the issuer then. If the serial number matches, that's good enough for this purpose.